### PR TITLE
Request: fixed cache for detectors and bug for __call() method

### PR DESCRIPTION
**Pull request by mirko-pagliai**
This fixes bug issue https://github.com/cakephp/cakephp/issues/9269, removing the cache for request detectors.

### CODE REVIEW

**Comment by markstory** [comment]
Looks good to me.

**Comment by ADmad** [comment]
Tests failing.

**Comment by lorenzo** [comment]
Should we just fix it? This will make things significantly slower

**Comment by mirko-pagliai** [comment]
@ADmad  that's a test that I not added, but that was already present. And that perhaps gave a positive result because the result was taken from the cache.

**Comment by mirko-pagliai** [comment]
@lorenzo, this is a bug. Previous tests were wrong.

This is the previous test:

        $request->params = ['controller' => 'cake'];
        $this->assertTrue($request->is('controller', 'cake'));
        $this->assertTrue($request->isController('cake'));

Try this test (added last line):

        $request->params = ['controller' => 'cake'];
        $this->assertTrue($request->is('controller', 'cake'));
        $this->assertTrue($request->isController('cake'));
        $this->assertTrue($request->is('controller', 'nonExistingController'));

I have not tried, but in logic the test should fail anyway, because it fetches the result `true` from the cache (the name of detectory is the cache key!).

Now, in your AppController try something like this:

    public function beforeFilter(\Cake\Event\Event $event) {
        if($this->request->is('controller', 'Posts')) {
            //do something for all actions of the Posts controller
        }

        //other code

 
        if($this->request->is('controller', 'Pages')) {
            //do something for all actions of the Pages controller
        }
    }

The second statement will have from the cache **always** the result of the first.

It is not a whim, my app has unpredictable behavior. This is a bug, man!
You can not take from the cache the value of something else!

**Comment by mirko-pagliai** [comment]
In other words:

   $this->request->is('mydetector', 'a value');
   $this->request->is('mydetector', 'another value');
   $this->request->is('mydetector', 'another value', 'another second value');

In the second and in the third case will be retrieved from the cache the result of the first case.

"Should we just fix it?". Yes, we should, it seems to me a several bug! Tell me if I'm wrong.

**Comment by josegonzalez** [comment]
@mirko-pagliai I think he means that the bug should be fixed rather than the caching dropped, not that this "isn't a problem".

**Comment by mirko-pagliai** [comment]
@lorenzo @josegonzalez so another way is: the cache must also consider the correspondence of the same arguments.

See https://github.com/cakephp/cakephp/issues/9269#issuecomment-239801835

**Comment by josegonzalez** [comment]
Correct. Would you be willing to update your pr to take that into account?

**Comment by mirko-pagliai** [comment]
There's a bug also on `__call()` method. It calls the `is()` method passing only the type name and not args.

Argh!

**Comment by mirko-pagliai** [comment]
This is the reason why the tests fail, there was already a bug present.

    $this->assertTrue($request->is('controller', 'cake'));
    $this->assertTrue($request->isController('cake'));

Due to the bug of the `__call()` method, in the second test the argument is not passed. But with the cache enabled, it still is positive, because it takes the previous test result.

Also the `is()` method is bugged, where the method is called recursively... argh!

**Comment by markstory** [comment]
One option might be to only use the cache when a detector has 0 additional arguments.

**Comment by mirko-pagliai** [comment]
So:
1) fixed bug on `__call()` method;
2) now I will restore the cache only if the method is called without additional arguments;
3) we have to fix the bug on https://github.com/cakephp/cakephp/blob/master/src/Network/Request.php#L639 (if called recursively, it is not passed additional arguments).

**Comment by mirko-pagliai** [comment]
The tests are positive.

There was a bug also in the `__call()` method. On the first test, the bug did not occur.
In next test, since it used the cache and took a wrong value, the bug has not been seen yet.

Removing the cache, the bug occurred on the tests that were already present prior to my request.

**Comment by markstory** [comment]
All [builds are currently failing](https://travis-ci.org/cakephp/cakephp/jobs/152484072) could that be fixed up?

File: src/Network/Request.php [hunk 1]
```diff
@@ -632,27 +634,31 @@ public function __isset($name)
...
+        array_shift($args);
 
         $type = strtolower($type);
         if (!isset(static::$_detectors[$type])) {
             return false;
         }
+        
+        if (empty($args)) {
+            if (!isset($this->_detectorCache[$type])) {
+                $this->_detectorCache[$type] = $this->_is($type);
```
Comment (markstory) [inline src/Network/Request.php]: Missing a second argument here.

Comment (mirko-pagliai) [inline src/Network/Request.php]: @markstory 'cause `$args` is empty :-)

**Comment by mirko-pagliai** [comment]
Ok, I will fix, np

File: src/Network/Request.php [hunk 2]
```diff
@@ -134,7 +134,7 @@ class Request implements ArrayAccess
         'json' => ['accept' => ['application/json'], 'param' => '_ext', 'value' => 'json'],
         'xml' => ['accept' => ['application/xml', 'text/xml'], 'param' => '_ext', 'value' => 'xml'],
     ];
-
+
```
Comment (dereuromark) [inline src/Network/Request.php]: Please make sure your IDE trims right. There should be no whitespace noise here.

File: src/Network/Request.php [hunk 3]
```diff
@@ -632,27 +634,31 @@ public function __isset($name)
...
+        
+        $args = func_get_args();
+        array_shift($args);
 
         $type = strtolower($type);
         if (!isset(static::$_detectors[$type])) {
             return false;
         }
+        
+        if (empty($args)) {
```
Comment (dereuromark) [inline src/Network/Request.php]: you could inverse to return early in the more simpler case:

```php
if ($args) {
    return $this->_is($type, $args);
}

...
```

File: src/Network/Request.php (hunk 2 above)
Comment (mirko-pagliai) [inline src/Network/Request.php]: You're right, damned Netbeans :-(

I apologize if I have to do all these PR, slowly I will become smarter.

**Comment by mirko-pagliai** [comment]
Fixed.

But note that the `is()` method can be called with no arguments, therefore the internal `_is()` method should be called with no arguments.

So maybe **(?)** it should fixed the method declaration, instead of method calls.

    protected function _is($type, array $args = [])

**Comment by mirko-pagliai** [comment]
Now it **should** be all right. I'm happy :-D

**Comment by lorenzo** [comment]
Thanks @mirko-pagliai

**Pull request merged**